Merge darwin/libdispatch-890 to master

das opened the pull request:
Merge tag darwin/libdispatch-890 on darwin/trunk

This includes the latest Apple source drop libdispatch-890 (up from libdispatch-806), on top of darwin/darwin-008 already merged in #252

das commented:
@swift-ci please test

das commented:
@swift-ci please test

das commented:
@swift-ci please test

das commented:
@swift-ci please test

das commented:
@swift-ci please test

dgrove-oss commented on src/swift/DispatchStubs.cc:
```diff
@@ -173,12 +173,6 @@ _swift_dispatch_retain(dispatch_object_t obj) {
-// _swift_apply_current_root_queue() {
-// 	return DISPATCH_APPLY_CURRENT_ROOT_QUEUE;
-// }
-
```
Should this also be taken out of Queue.swift in the overlay?  There's a silgen directive for this symbol towards the bottom of that file.

das replied:
yup, good point, thanks!
fixed

das commented:
@dgrove-oss  thanks! removed that obsolete symbol in b9b44c3

would you mind giving this branch a local sanity check on your setup? I fixed all the build errors that the CI found but it hit some weird [errors](https://ci.swift.org/job/swift-corelibs-libdispatch-PR-Linux/105/) that I think are unrelated but it would be good to be sure, given the size of the change and how close we are to the next 4.0 tag.

FWIW my plan is to merge master back into swift-4.0 branch after we have merged this PR

das commented:
@swift-ci please test

das commented:
@swift-ci please test

dgrove-oss commented:
It's crashing for me on pretty much every dispatch test that uses non-overcommit queues.  Here's the backtrace.  I'd assume it is something relatively simple, but I might not have a chance to dig into it until later tonight.  Have to leave work a bit early today.

```
714			DISPATCH_INTERNAL_CRASH(0, "Lingering DISPATCH_WLH_ANON");
(gdb) bt
#0  _dispatch_adopt_wlh_anon () at ./inline_internal.h:714
#1  _dispatch_mgr_thread (dq=0x7ffff7ff0fc0 <_dispatch_mgr_q>, dic=0x7ffff5d37d18, flags=196608) at source.c:2324
#2  0x00007ffff7fa3c75 in _dispatch_continuation_pop_inline (dou=..., dic=0x7ffff5d37d18, flags=196608, dq=0x7ffff7ff0f40 <_dispatch_mgr_root_queue>)
    at ./inline_internal.h:2505
#3  0x00007ffff7fafe8a in _dispatch_root_queue_drain (dq=0x7ffff7ff0f40 <_dispatch_mgr_root_queue>, pp=0) at queue.c:5710
#4  0x00007ffff7faf92d in _dispatch_worker_thread (context=0x7ffff7ff0f40 <_dispatch_mgr_root_queue>) at queue.c:5819
#5  0x00007ffff79ac6ba in start_thread (arg=0x7ffff5d38700) at pthread_create.c:333
#6  0x00007ffff76e282d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
```

das commented:
ugh, thanks, we should be able to figure that one out by source inspection

das commented:
@dgrove-oss	7de6bba should fix it, please have another go, thanks!

das commented:
@swift-ci please test

das commented:
@swift-ci please test

das commented:
looks like the CI was happy this time, so the previous failures may have been (non-obviously) a consequence of that crash

dgrove-oss commented:
Confirmed.  The test suite runs cleanly for me as well.

das commented:
awesome, thanks for confirming Dave!

I'll squash and merge
